zeebe/engine/src/test/java/io/camunda/zeebe/engine/processing/bpmn/activity/listeners/task/TaskListenerTest.java

```diff
@@ -0,0 +1,496 @@
+    final long processInstanceKey =
+        createProcessInstance(
+            createProcess(
+                p ->
+                    p.userTask(
+                            USER_TASK_ELEMENT_ID,
+                            t ->
+                                t.zeebeUserTask()
+                                    .zeebeAssignee("foo")
+                                    .zeebeTaskListener(l -> l.complete().type(LISTENER_TYPE))
+                                    .zeebeOutput("=my_listener_var+\"_abc\"", "userTaskOutput"))
+                        .serviceTask(
+                            "subsequent_service_task",
+                            tb -> tb.zeebeJobType("subsequent_service_task"))));
+
+    ENGINE.userTask().ofInstance(processInstanceKey).complete();
+
+    // when: complete TL job with a variable 'my_listener_var'
+    ENGINE
+        .job()
+        .ofInstance(processInstanceKey)
+        .withType(LISTENER_TYPE)
+        .withVariable("my_listener_var", "bar")
+        .complete();
+
+    // then: assert the variable 'userTaskOutput' is accessible in the subsequent element
+    final var subsequentServiceTaskJob = activateJob(processInstanceKey, "subsequent_service_task");
+    assertThat(subsequentServiceTaskJob.getVariables()).containsEntry("userTaskOutput", "bar_abc");
+    completeJobs(processInstanceKey, "subsequent_service_task");
+  }
+
+  @Test
+  public void shouldIncludeConfiguredUserTaskDataInCompleteTaskListenerJobHeaders() {
+    final var form = deployForm("/form/test-form-1.form");
+    final long processInstanceKey =
+        createProcessInstance(
+            createProcessWithZeebeUserTask(
+                t ->
+                    t.zeebeAssignee("admin")
+                        .zeebeCandidateUsers("user_A, user_B")
+                        .zeebeCandidateGroups("group_A, group_C, group_F")
+                        .zeebeFormId("Form_0w7r08e")
+                        .zeebeDueDate("2095-09-18T10:31:10+02:00")
+                        .zeebeTaskListener(l -> l.complete().type(LISTENER_TYPE))));
+
+    // when
+    final var userTaskRecordValue = ENGINE.userTask().ofInstance(processInstanceKey).complete();
+
+    // then
+    final var activatedListenerJob = activateJob(processInstanceKey, LISTENER_TYPE);
+
+    assertThat(activatedListenerJob.getCustomHeaders())
+        .containsOnly(
+            entry(
+                Protocol.USER_TASK_CANDIDATE_GROUPS_HEADER_NAME,
+                "[\"group_A\",\"group_C\",\"group_F\"]"),
+            entry(Protocol.USER_TASK_CANDIDATE_USERS_HEADER_NAME, "[\"user_A\",\"user_B\"]"),
+            entry(Protocol.USER_TASK_ASSIGNEE_HEADER_NAME, "admin"),
+            entry(Protocol.USER_TASK_DUE_DATE_HEADER_NAME, "2095-09-18T10:31:10+02:00"),
+            entry(Protocol.USER_TASK_FORM_KEY_HEADER_NAME, Objects.toString(form.getFormKey())),
+            entry(USER_TASK_KEY_HEADER_NAME, String.valueOf(userTaskRecordValue.getKey())));
```

Comment:
❓ Why is the `USER_TASK_KEY_HEADER_NAME` not added to `Protocol`?

Reply:
I'm not sure that we have a strong decision yet on whether `USER_TASK_KEY_HEADER_NAME` should be part of the custom headers. Or it could be included as a separate property in the `JobRecord`.
Once that's decided, we can make it part of the public `Protocol` API.